### PR TITLE
OCM-3313 | refactor: extract autoscaler code to new modules

### PR DESCRIPTION
**Pull request by osherdp**
In order to make code less repetitive when we add day2 support of cluster-autoscaler (and also to reduce the amount of code in ``cmd/create/cluster/cmd.go``), this refactors the code to be placed in new modules that are also more easily imported by new commands.

In the step 2 of this we want to add a ``rosa create autoscaler`` command that will use the new location of the code

### CODE REVIEW

**Comment by osherdp** [comment]
/hold
waiting for reviews

**Comment by vkareh** [comment]
@osherdp Do you have a ticket and/or DDR for this change? I have some reservations about introducing a `create` command for something that's not a resource in our API. Would `rosa edit cluster` be a better command to tackle this?

**Comment by gdbranco** [comment]
Mostly moving code around, SGTM

File: pkg/ocm/validators.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package ocm
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+func IntValidator(val interface{}) error {
```
Comment (ciaranRoche) [inline pkg/ocm/validators.go:9]: Can we add unit tests for all these validators :pray:

**Comment by vkareh** [comment]
@osherdp but considering that this is also a day-1 option, it seems more fitting for `rosa create|edit cluster`. Sounds like a mistake to have put it in a separate API endpoint and should be part of `Cluster.AutoScaler{}`?

This code change means users can no longer set it during cluster creation, so it seems we're changing/removing functionality

**Comment by osherdp** [comment]
> but considering that this is also a day-1 option, it seems more fitting for rosa create|edit cluster. Sounds like a mistake to have put it in a separate API endpoint and should be part of Cluster.AutoScaler{}?

Maybe we can incorporate that into ``rosa edit cluster`` rather than ``rosa create|edit autoscaler``, but either way it's out of scope for this specific PR. Whatever the command we choose we prefer to have the code re-usable across the different commands (instead of repeating a lot of code)

The reason I think ``rosa create|edit autoscaler`` is more clear is that it's more consistent with the existing behavior of default machine-pool | oidc-providers. I don't think we can justify those being separate resources and autoscaler being a cluster property

> This code change means users can no longer set it during cluster creation, so it seems we're changing/removing functionality

after this code change users are still able to set it during cluster creation. it's only a preparation for handling it in a day2 scenario

**Comment by vkareh** [comment]
okay, I see your point, thanks for clarifying. So basically customers can set it during `rosa create cluster` as well as `rosa create autoscaler` if I understand correctly? If that's the case, then I'm fine with it. Just want to make sure it doesn't cause confusion to end-users. Since day-1 `create cluster` resources are perceived as part of the cluster spec, users will likely expect them to also be editable as part of the cluster spec (which implies `edit cluster`). Does that make sense?

**Comment by osherdp** [comment]
> 

@vkareh yeah, I get what you mean. Actually originally I also wanted to make it all part of the cluster object, but we've got into a situation where the lack of support for ``PUT`` requests in OCM made it impossible to partially override autoscaler configuration and still preserve the ``PATCH`` contract. Because requirements were to allow that for day1 and day2, we didn't have much choice and had to have it this way

In this case we partly rely on the docs making this kind of behavior clearer to users. Either way configuring cluster-autoscaler is something I consider a bit more advance, so I assume people will first turn to the docs and won't expect it to be intuitive on first sight. Hope it makes some sense :smile:

**Comment by vkareh** [comment]
@osherdp yes! Our API unfortunately doesn't have a way to unset attributes, apart from issuing a `DELETE` request. Sigh...

Okay then, this all makes more sense to me. Thanks for all this context! :+1:

**Comment by osherdp** [comment]
@ciaranRoche @gdbranco please let me know if anything else is needed
Tests are added to the validators and this should only refactor the code (we still have the next step in order to introduce the new commands)

**Comment by osherdp** [comment]
/hold
sorry, doing one more manual test

**Review by ciaranRoche — APPROVED** [review]
/lgtm

**Comment by osherdp** [comment]
/unhold

thanks everyone!

**Pull request merged**